### ISSUE TITLE
`RandomLinkSplit` enhancement for undirected graphs

### ISSUE THREAD

**Issue by saiden89** [pyg-team/pytorch_geometric#5190]
### 🚀 The feature, motivation and pitch

As briefly discussed in #5169, `RandomLinkSplit` currently returns `edge_label_index` in directed form, regardless of the input graph being undirected or the arguments passxfed to the transform. We can extend to possibility to return appropriate supervision edges according to the usecase.
As far as I can tell we can do so in various ways:
- Let PyG guess the directiveness of the input graph  and perform the split accordingly: this way undirected graphs will have only undirected edge indices and viceversa. Additionally, give the user the possibility to override the settings.
- Keep the current defaults (without guessing), but give the user the possibility to override.

As far as my (quite limited) experience goes, if the usecase is centered on undirected network(s) we will probably stick to those throughout all the steps. The same goes for the directed counterpart. Even when dealing with heterogenous network, there is rarely the need to have different types of message passing (MP) and supervision edges within the same edge types. So I guess that for most cases keeping in sync MP and validation indices is preferrable.

### Alternatives

Currently, it is possible to manually carry out the necessary transformations after the split, on a per use basis.

**Comment by rusty1s** [comment]
I have no strong opinion here. As discussed in #5169, I think that if one wants to predict undirected edges, one should use a symmetric encoder or take care of both directions on its own. We can make this more clear in the documentation.

Wondering if other people have some thoughts here (@Padarn @lightaime @wsad1) - if this feels indeed overly complicated, then let's change it :)

**Comment by Padarn** [comment]
> I think that if one wants to predict undirected edges, one should use a symmetric encoder or take care of both directions on its own.

I tend to agree with this. But I do think the current interface makes this a bit unclear (its a little confusing that the `undirected` flag does not make the labels undirected). I'd prefer to provide clear documentation and leave it as is.

But also not a very strong opinion.

**Comment by Tigerrr07** [comment]
# Problem background
When applying `RandomLinkSplit` in an undirected HeteroData `data`, in figure1 (I already store reverse edges in original `edge_index`, see figure 2 as an example), I find  in splitted data, such as `train_data`, `val_data` or `test_data`, the `edge_index` of one `edge_type` is not undirected , I test it using `is_undirected()` method.
![image](https://user-images.githubusercontent.com/86994637/184830074-edfb2210-7e02-4d1d-9ae7-6ec860bfe47e.png)
![image](https://user-images.githubusercontent.com/86994637/184831326-939fcf8b-c4dd-45b9-8cb5-9ac4b6dc379d.png)

# Code
``` python
node_types, edge_types = data.node_types, data.edge_types
# split data
train_data, val_data, test_data = T.RandomLinkSplit(
        num_val=0.1,
        num_test=0.1,
        is_undirected=True,
        add_negative_train_samples=False,
        disjoint_train_ratio=0,
        edge_types=edge_types,
        rev_edge_types=edge_types
    )(data)

# test if undirected
from torch_geometric.utils import is_undirected

edge_type = ('drug', 'anaemia','drug')
print(len(data[edge_type].edge_index[0]), is_undirected(data[edge_type].edge_index)) # 54012, True
print(len(train_data[edge_type].edge_index[0]), is_undirected(train_data[edge_type].edge_index)) # 43210, False
print(len(val_data[edge_type].edge_index[0]), is_undirected(val_data[edge_type].edge_index))   # 43210, False
print(len(test_data[edge_type].edge_index[0]), is_undirected(test_data[edge_type].edge_index))  # 48611, False
```
# Problem
* The number of edges in `test_data` is odd, but this should be an **even** number related to bidirectional edges in undirected graphs.
* `edge_index` in `train_data, val_data, test_data` should be undirected.

In homogeneous graphs, there are no such problems. But I don't know why this happen in my constructed heterogeneous graph.

**Comment by rusty1s** [comment]
See https://github.com/pyg-team/pytorch_geometric/issues/5206
